### DISCUSSION TITLE
Thanks for papermerge, but why its so hard to install??

### DISCUSSION THREAD

**Discussion by kolossboss** [papermerge/papermerge-core#270] [General]
Hi,

First of all, thanks for this amazing product!! And the Documentation is also great.
I'm running Papermerge for a few month with the Image from "LinuxServer.io" release group. It was really easy to install and runs perfect. But the last working release has a few annoying bugs and the Image is not updated.  (Its deprecated)

So I came here for the official image.
But I have a very hard time to set this up.
I tried via Docker Compose and with Portainer and repo.

The main Issue is, I always get the "404 not found" Error when accessing the WebUI  [(same as this Issue)](https://github.com/papermerge/papermerge-core/issues/17).
Its think because of that "HOSTNAME" or "USE_HOSTNAME"  variable.
When I use the Domain Name I get this 404 Error. When Using the IP-Adress of the Host, I get the Login UI, but I also get this Syntax Error....

**I have given up..**

I don't know if I'm just stupid, but I also run over 20 other Docker Containers (not all from LinuxServer) with great success.
I'm not that deep into "creating Containers" but why is the Image from LinuxServer so simple and easy to use and also possible with one Image, and this official image gives me a headache.

I will stay a little longer on the LinuxServer image, and hope for a fix or anything. I'm not sure.

Best regards

Simon

**Comment by ciur** [comment]
@kolossboss, thanks for opening this ticket.

Indeed, the latest release 2.1.x is a pain in the ass to install. One of difficulties is the fact that frontend and backend are separate applications and in order to have basic backend + frontend you need to have those two docker containers plus some nginx/traefik in between. 
Now, the idea of having backend and frontend as separate apps, is very cool, modern and helps me to have a clean and easy to maintain code (separation of concerns etc).
However, as you, and many others noted - it is a pain to configure.

Long story short, yesterday, i've decided to change that, in other words to make "one docker image for both backend and frontend", similar to what it used to be for 2.0.x (and what linuxserver.io image offers for 2.0.x).
The end result of this is that instead of 3 docker images to run basic app, you will need just one.

Actually you can already try that:
```
docker run -p 9090:8000   -e PAPERMERGE__MAIN__SECRET_KEY=abc \              
    -e DJANGO_SUPERUSER_PASSWORD=123 \
    papermerge/papermerge:2.1.8.dev8
```

then just point you browser to http://localhost:9090, username=admin, password=123, and you will see (and can  use) both the UI and REST API!

Note that it is development image  `2.1.8.dev8`, i.e. not yet officially released, it is only for me to play around (still it is available on docker hub: https://hub.docker.com/r/papermerge/papermerge/tags).

Here is the official ticket tracking the progress on this problem: https://github.com/ciur/papermerge/issues/537
I need max a week or so to get it done (+ update docs etc).

Btw, there is yet another, less obvious, but important difference between 2.0.x (provided by LinuxServer) and 2.1.x.
2.0.x - used postgresql built int search engine for searches. 
2.1.x - uses xapian as built in search engine, but the recommendation is to use elasticsearch (or solr) instead.
This feat adds complexity indeed, but I think it is worth it, as in long term I will be able to fine tune the searching part so that
search results (one of the major points of Papermerge) will be more to the point. 


Last but not least, is the worker, which I think LinuxServer docker image runs it as part of the same docker container. 
As I am pondering over that idea, I think it makes sense. Also I gained more xp in docker containers matters :) In any case, for near future (couple of weeks or so) I will provide
a docker image tagged "bundle" e.g papermerge/papermerge:2.1.9-bundle which will run one restapi server, worker websockets server and web UI **in one single docker imag**e. This will help people who want to stick with simple and quick setups (linux linuxserver provides for 2.0.x).

**Comment by kolossboss** [comment]
@ciur 
Thanks for the kind reply.
And even bigger thanks for the very detailed explanation!

Now I understand the why you did this. And it was probably the better way from a developer's point of view.👍 
I will definitely try the development Image. And I'm looking forward for the new "single container" release.
Thank you for taking user problems so seriously!

**Comment by ciur** [comment]
@kolossboss, [here](https://github.com/papermerge/papermerge-core/blob/master/docker/docker-compose.yml) is new, simplified docker compose file.
Notice there is nothing about traefik, frontend/backend, web sockets in it - now there are just 4 services which, I guess, should make perfect sense to have around:
 -  rest api backend
 -  worker (for performing OCR)
 - db 
 - redis

Now, the simplest way of lunching papermerge:
```
  docker run -p 16000:8000 \
      -e PAPERMERGE__MAIN__SECRET_KEY=abc \
      -e DJANGO_SUPERUSER_PASSWORD=123 \
      papermerge/papermerge:2.1.8
```
will include UI as well, on other words, if you point your browser to http://localhost:16000, you will see papermege login window. 

If you experience any "gray screen", please either delete cookies for that domain (localhost:16000) or start in incognito mode.

**Comment by kolossboss** [comment]
Thanks a lot.
A few days ago, I tried the "dev" Version and its working fine.
Now I get this error when I lunching Papermerge via Terminal:

`docker: no matching manifest for linux/arm64/v8 in the manifest list entries.`

And this via compose and Docker

`failed to deploy a stack: redis Pulling backend Pulling worker Pulling db Pulling redis Error backend Error no matching manifest for linux/arm64/v8 in the manifest list entries`

I'm running Docker on Apple Silicon M2.

**Comment by ciur** [comment]
@kolossboss, can you try it again, please?

Just make sure you (re) fetch (pull again) 2.1.8 docker image - as I recreated it.

 By mistake I removed "building of docker images for arm64"  platforms. I added that part back and [recreated 2.1.8 docker image](https://hub.docker.com/r/papermerge/papermerge/tags). Now docker image (2.1.8) should run on M2 processors (which are arm64) as well.

**Comment by kolossboss** [comment]
@ciur 
Now its working! Thanks.

But this is still a test version right? I missing features like importing or automates.

**Comment by mblahay** [comment]
I too would like to know whether 2.1 is still in test phase.

**Comment by ciur** [comment]
The next release is 3.0, which is now in testing phase.

You can try it:

- via docker: https://papermerge.github.io/documentation/3.0-dev/setup/docker/
- via docker compose: https://papermerge.github.io/documentation/3.0-dev/setup/docker-compose/
- or ansible: https://papermerge.github.io/documentation/3.0-dev/setup/ansible/

I have simplified the setup a lot. Now it is easier start with Papermerge.

**Reply by Narrat** [reply]
Thanks for all the work.
Trying the 3.0 release I stumble onto this, when trying to build this in a fakeroot/venv like environment:
```
validate_pyproject.api.load_builtin_plugin defines `tool.distutils` schema
validate_pyproject.api.load_builtin_plugin defines `tool.setuptools` schema
Traceback (most recent call last):
  File "/usr/lib/python3.11/site-packages/pyproject_hooks/_in_process/_in_process.py", line 353, in <module>
    main()
  File "/usr/lib/python3.11/site-packages/pyproject_hooks/_in_process/_in_process.py", line 335, in main
    json_out['return_val'] = hook(**hook_input['kwargs'])
                             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/lib/python3.11/site-packages/pyproject_hooks/_in_process/_in_process.py", line 118, in get_requires_for_build_wheel
    return hook(config_settings)
           ^^^^^^^^^^^^^^^^^^^^^
  File "/usr/lib/python3.11/site-packages/setuptools/build_meta.py", line 325, in get_requires_for_build_wheel
    return self._get_build_requires(config_settings, requirements=['wheel'])
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/lib/python3.11/site-packages/setuptools/build_meta.py", line 295, in _get_build_requires
    self.run_setup()
  File "/usr/lib/python3.11/site-packages/setuptools/build_meta.py", line 311, in run_setup
    exec(code, locals())
  File "<string>", line 1, in <module>
  File "/usr/lib/python3.11/site-packages/setuptools/__init__.py", line 103, in setup
    return distutils.core.setup(**attrs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/lib/python3.11/site-packages/setuptools/_distutils/core.py", line 159, in setup
    dist.parse_config_files()
  File "/usr/lib/python3.11/site-packages/setuptools/dist.py", line 627, in parse_config_files
    pyprojecttoml.apply_configuration(self, filename, ignore_option_errors)
  File "/usr/lib/python3.11/site-packages/setuptools/config/pyprojecttoml.py", line 72, in apply_configuration
    config = read_configuration(filepath, True, ignore_option_errors, dist)
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/lib/python3.11/site-packages/setuptools/config/pyprojecttoml.py", line 133, in read_configuration
    validate(subset, filepath)
  File "/usr/lib/python3.11/site-packages/setuptools/config/pyprojecttoml.py", line 61, in validate
    raise ValueError(f"{error}\n{summary}") from None
ValueError: invalid pyproject.toml config: `project`.
configuration error: `project` must contain ['name'] properties

ERROR Backend subprocess exited when trying to invoke get_requires_for_build_wheel
```
And I puzzled by this error message. Any idea or hint about the cause?

**Reply by ciur** [reply]
How did you start (i.e. what command you used) the project ?
Could you please provide details about " build this in a fakeroot/venv" part ?

**Reply by Narrat** [reply]
Sure.
In general I'm trying to revive the papermerge PKGBUILD which resides in the user repository ( https://aur.archlinux.org/packages/papermerge / https://aur.archlinux.org/packages/papermerge-core ). So classic package distribution at work.
The tool for creating the package (`makepkg`) creates a fakeroot where the source is built.
In and itself I'm trying to build the package via PEP 517 standard methods via the source tarball from the release.
Package versions for the build process: `build 1.0.3`, `installer 0.7.0`, `wheel 0.42.0`, `poetry-core 1.8.1` & `setuptools 69.0.3`.
Command at work: `python -m build --wheel --no-isolation`
Even if there are build deps missing it should fail with something like "couldn't import xyz". And it does find the `pyproject.toml`. But I'm at a loss why it fails to parse the file. In general you use the same structure with subcomponents (for example salinic) and there it works without a problem.

Edit in 2025:
The error message was on point. The `pyproject.toml` was not PEP-517 compliant. Dunno why I didn't see the missing section.
I saw that the situation will improve with the upcoming `3.4` release, which includes #580 . But it still doesn't seem to be PEP-517 compliant.

**Comment by kolossboss** [comment]
Hi thanks,
I tried v3 on my M2 MacMini but there is no image for linux/arm64.

I also tried to emulate arm64 by adding   `platform: linux/amd64/v8` to the compose file.
All container now starts successfully, I can now login to papermerge. But then I get `502 Bad Gateway`.
Also my CPU is running at 100% for 30minutes.

**Reply by ciur** [reply]
There is no papermerge dockerimage `arm` support at the moment. I used to provide one, but I've decided to discontinue `arm`support, for the simple reason that I don't own any `arm` device and there is no way for me to test it. I am looking forward to the community to contribute/build/test `arm` images, which should not be a difficult task given the fact that `linux/amd64` images are already out there.
